sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobInputOutputStreamTest.groovy

```diff
@@ -147,30 +148,308 @@ class BlockBlobInputOutputStreamTest extends APISpec {
+
+        then: "Successful read"
+        def b
+        while ((b = inputStream.read()) != -1) {
+            outputStream.write(b)
+        }
+
+        byte[] randomBytes1 = outputStream.toByteArray()
+        assert randomBytes1 == randomBytes
+    }
+
+    // Error case
+    def "IS consistent read control etag user provides old etag"() {
+        setup:
+        int length = Constants.KB
+        byte[] randomBytes = getRandomByteArray(length)
+        def blobContainerClient = versionedBlobServiceClient.createBlobContainer(containerName)
+        def blobClient = blobContainerClient.getBlobClient(generateBlobName())
+        blobClient.upload(new ByteArrayInputStream(randomBytes), length, true)
+        def properties = blobClient.getProperties()
+
+        when:
+        def inputStream = blobClient.openInputStream(new BlobInputStreamOptions().setBlockSize(1).setConsistentReadControl(ConsistentReadControl.ETAG)
+            .setRequestConditions(new BlobRequestConditions().setIfMatch(properties.getETag())))
+
+        // Since eTag is the only form of consistentReadControl and the blob is modified, we will throw.
+        blobClient.upload(new ByteArrayInputStream(randomBytes), length, true)
+
+        inputStream.read()
+
+        then: "Failed read"
+        thrown(IOException) // BlobStorageException = ConditionNotMet
+    }
+
+    def "IS consistent read control version client chooses version"() {
+        setup:
+        int length = Constants.KB
+        byte[] randomBytes = getRandomByteArray(length)
+        def blobContainerClient = versionedBlobServiceClient.createBlobContainer(containerName)
+        def blobClient = blobContainerClient.getBlobClient(generateBlobName())
+        blobClient.upload(new ByteArrayInputStream(randomBytes), length, true)
+
+        when:
+        // No version specified - client will lock on it.
+        def inputStream = blobClient.openInputStream(new BlobInputStreamOptions().setConsistentReadControl(ConsistentReadControl.VERSION_ID))
+        def outputStream = new ByteArrayOutputStream()
+
+        // When a versioned client is used it should still succeed if the blob has been modified
+        blobClient.upload(new ByteArrayInputStream(getRandomByteArray(length)), length, true)
+
+        then:
+        def b
+        while ((b = inputStream.read()) != -1) {
+            outputStream.write(b)
+        }
+        byte[] randomBytes2 = outputStream.toByteArray()
+        assert randomBytes2 == randomBytes
+    }
+
+    def "IS consistent read control version user provides version"() {
+        setup:
+        int length = Constants.KB
+        byte[] randomBytes = getRandomByteArray(length)
+        def blobContainerClient = versionedBlobServiceClient.createBlobContainer(containerName)
+        def blobClient = blobContainerClient.getBlobClient(generateBlobName())
+        blobClient.upload(new ByteArrayInputStream(randomBytes), length, true)
+        def properties = blobClient.getProperties()
+
```

Comment:
Can we do another upload after we get the properties to force this version to be different from the latest at the time of constructing the stream?